### ISSUE TITLE
SIGBUS in __pool_get_from_cache()

### ISSUE THREAD

**Issue by jpulz** [haproxy/haproxy#166]
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.0.1 2019/06/26 - https://haproxy.org/
Build options :
  TARGET  = freebsd
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -pipe -march=ivybridge -fstack-protector-strong -fno-strict-aliasing -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-ignored-qualifiers -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wtype-limits -Wshift-negative-value -Wnull-dereference -DFREEBSD_PORTS
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_REGPARM=1 USE_STATIC_PCRE=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_ACCEPT4=1 USE_ZLIB=1 USE_CPU_AFFINITY=1

Feature list : -EPOLL +KQUEUE -MY_EPOLL -MY_SPLICE -NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM +STATIC_PCRE -STATIC_PCRE2 +TPROXY -LINUX_TPROXY -LINUX_SPLICE +LIBCRYPT -CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL -LUA -FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY -TFO -NS -DL -RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER -PRCTL -THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1a-freebsd  20 Nov 2018
Running on OpenSSL version : OpenSSL 1.1.1a-freebsd  20 Nov 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with transparent proxy support using: IP_BINDANY IPV6_BINDANY
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE version : 8.43 2019-02-23
Running on PCRE version : 8.43 2019-02-23
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes

Available polling systems :
     kqueue : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use kqueue.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE        mux=H2
              h2 : mode=HTX        side=FE|BE     mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services : none

Available filters :
        [SPOE] spoe
        [COMP] compression
        [CACHE] cache
        [TRACE] trace
```

```
FreeBSD lb02.local 12.0-RELEASE-p6 FreeBSD 12.0-RELEASE-p6 r349363 LOADBALANCER amd64
```

## What's the configuration?

```
global
  stats socket    /var/run/haproxy.stat mode 600 level admin
  log             /var/run/log local0 info
  user            www
  group           www
  daemon
  node lb02
  ssl-default-bind-options ssl-min-ver TLSv1.2
  tune.ssl.default-dh-param 2048

defaults
  log global
  mode http
  option httplog
  option dontlognull
  timeout connect 60000ms
  timeout client 30000ms
  timeout server 60000ms
  timeout check 60000ms

listen fe_stats_https
  mode http
  option forwardfor except 127.0.0.1
  bind 192.168.0.1:80
  bind 192.168.0.1:443 ssl crt /usr/local/etc/ssl/cert.pem
  redirect scheme https code 301 if !{ ssl_fc }   # redirect 80 -> 443 (for owa)
  acl stats url_beg /stats
  use_backend be_stats if stats

backend be_stats
  mode http
  stats enable
  stats uri /stats
  stats refresh 15s
  stats show-desc EXLB02
  stats show-legends
  stats show-node

frontend fe_ex2016_https
  mode http
  option forwardfor except 127.0.0.1
  bind 192.168.0.12:80
  bind 192.168.0.12:443 ssl crt /usr/local/etc/ssl/cert.pem
  redirect scheme https code 301 if !{ ssl_fc }   # redirect 80 -> 443 (for owa)
  acl autodiscover url_beg /autodiscover
  acl autodiscover url_beg /Autodiscover
  acl mapi url_beg /mapi
  acl rpc url_beg /rpc
  acl owa url_beg /owa
  acl owa url_beg /OWA
  acl eas url_beg /microsoft-server-activesync
  acl eas url_beg /Microsoft-Server-ActiveSync
  acl ecp url_beg /ecp
  acl ews url_beg /ews
  acl ews url_beg /EWS
  acl oab url_beg /oab
  acl oab url_beg /OAB
  use_backend be_ex2016_autodiscover if autodiscover
  use_backend be_ex2016_mapi if mapi
  use_backend be_ex2016_rpc if rpc
  use_backend be_ex2016_owa if owa
  use_backend be_ex2016_eas if eas
  use_backend be_ex2016_ecp if ecp
  use_backend be_ex2016_ews if ews
  use_backend be_ex2016_oab if oab
  default_backend be_ex2016

backend be_ex2016_autodiscover
  mode http
  balance roundrobin
  option httpchk GET /autodiscover/healthcheck.htm
  option log-health-checks
  http-check expect status 200
  server 192.168.0.2 192.168.0.2:443 check ssl inter 15s verify required ca-file /etc/ssl/cert.pem
  server 192.168.0.3 192.168.0.3:443 check ssl inter 15s verify required ca-file /etc/ssl/cert.pem

# skip - as other backends (for mapi, rpc, owa, eas, ecp, ews, oab) are nearly identical, only the httpchk URL is different

backend be_ex2016
  mode http
  balance roundrobin
  server 192.168.0.2 192.168.0.2:443 check ssl inter 15s verify required ca-file /etc/ssl/cert.pem
  server 192.168.0.3 192.168.0.3:443 check ssl inter 15s verify required ca-file /etc/ssl/cert.pem

##### TCP

frontend fe_ex2016_smtp
  bind 192.168.0.12:25 name SMTP
  mode tcp
  option dontlognull
  maxconn 10000
  timeout client 300000
  option tcplog
  option contstats
  option socket-stats
  default_backend be_ex2016_smtp

backend be_ex2016_smtp
  mode tcp
  balance leastconn
  option tcp-check
  tcp-check connect port 25
  tcp-check expect rstring "^220 "
  tcp-check send "EHLO exlb02.local\r\n"
  tcp-check expect rstring "^250-"
  tcp-check send "MAIL FROM: <>\r\n"
  tcp-check expect rstring "^250 "
  tcp-check send "QUIT\r\n"
  tcp-check expect rstring "^221 "
  option log-health-checks
  default-server on-marked-down shutdown-sessions
  server 192.168.0.2 192.168.0.2:25 check verify none
  server 192.168.0.3 192.168.0.3:25 check verify none

frontend fe_imap
  bind 192.168.0.12:143 name IMAP
  mode tcp
  option dontlognull
  maxconn 10000
  timeout client 300000
  option tcplog
  option contstats
  option socket-stats
  default_backend be_imap 

backend be_imap 
  mode tcp
  balance leastconn
  option tcp-check
  tcp-check connect port 143
  tcp-check expect string *\ OK
  tcp-check send "1 LOGOUT"
  option log-health-checks
  default-server on-marked-down shutdown-sessions
  server 192.168.0.2 192.168.0.2:143 check verify none
  server 192.168.0.3 192.168.0.3:143 check verify none

# skip other tcp-checks as they are nearly identical
```

## Steps to reproduce the behavior

Not known. It just occured once and only on one host (virtual machine) today.

## Actual behavior

haproxy stopped with SIGBUS in __pool_get_from_cache()

The system is a completely idle backup host for redundancy. Just the backend checks to Exchange 2016 servers are running with various backends going up and down. No client connections to haproxy at all.

From a quick look at the debugger (see output below) it seems to be related to tcp-check, but I'm not sure.

```
* thread #1, name = 'haproxy', stop reason = signal SIGBUS: hardware error
    frame #0: 0x0000000000377b9e haproxy`__pool_get_from_cache(pool=0x000000000066d480) at memory.h:199
   196          ph->count--;
   197          pool_cache_bytes -= ph->size;
   198          pool_cache_count--;
-> 199          LIST_DEL(&item->by_pool);
   200          LIST_DEL(&item->by_lru);
   201  #ifdef DEBUG_MEMORY_POOLS
   202          /* keep track of where the element was allocated from */
```
```
(lldb) bt
* thread #1, name = 'haproxy', stop reason = signal SIGBUS: hardware error
  * frame #0: 0x0000000000377b9e haproxy`__pool_get_from_cache(pool=0x000000000066d480) at memory.h:199
    frame #1: 0x00000000003779c0 haproxy`__pool_get_first(pool=0x000000000066d480) at memory.h:216
    frame #2: 0x0000000000377975 haproxy`pool_alloc_dirty(pool=0x000000000066d480) at memory.h:259
    frame #3: 0x00000000003777f5 haproxy`pool_alloc(pool=0x000000000066d480) at memory.h:273
    frame #4: 0x0000000000377d05 haproxy`conn_new at connection.h:499
    frame #5: 0x0000000000374e65 haproxy`cs_new(conn=0x0000000000000000) at connection.h:530
    frame #6: 0x0000000000372b48 haproxy`tcpcheck_main(check=0x0000000801542ec0) at checks.c:2817
    frame #7: 0x000000000037bcd3 haproxy`connect_conn_chk(t=0x0000000802695780) at checks.c:1596
    frame #8: 0x000000000037a615 haproxy`process_chk_conn(t=0x0000000802695780, context=0x0000000801542ec0, state=516) at checks.c:2204
    frame #9: 0x0000000000379dbe haproxy`process_chk(t=0x0000000802695780, context=0x0000000801542ec0, state=516) at checks.c:2375
    frame #10: 0x000000000047d7f5 haproxy`process_runnable_tasks at task.c:414
    frame #11: 0x0000000000389262 haproxy`run_poll_loop at haproxy.c:2516
    frame #12: 0x0000000000386f51 haproxy`run_thread_poll_loop(data=0x0000000000000000) at haproxy.c:2637
    frame #13: 0x00000000003834c5 haproxy`main(argc=4, argv=0x00007fffffffeaa8) at haproxy.c:3311
    frame #14: 0x00000000002a611b haproxy`_start(ap=<unavailable>, cleanup=<unavailable>) at crt1.c:76
```
```
(lldb) p *ph
(pool_cache_head) $1 = {
  list = {
    n = 0x0000000808523700
    p = 0x00000008085277c0
  }
  size = 416
  count = 200
}
```
```
(lldb) p *item
(pool_cache_item) $2 = {
  by_pool = {
    n = 0x02dc20000ef4c380
    p = 0x000000000055ac00
  }
  by_lru = {
    n = 0x0000000808565d10
    p = 0x0000000800db0350
  }
}
```

## Expected behavior

No SIGBUS at all.

## Do you have any idea what may have caused this?

No.

## Do you have an idea how to solve the issue?

No.

## Note

I can provide more debugging output as the process is still hanging in there at the current state.

Thanks

**Comment by cognet** [comment]
Hi,

Any chance you can checkout the latest 2.0 version from git, some commits that could fix that have been pushed, since the 2.0.1 release.

Thanks !

**Comment by jpulz** [comment]
Done.
Running on two hosts now.

Will report back as soon as anything happens.

Thanks

**Comment by jpulz** [comment]
Until now, no more appearances for this error so it seems to be already fixed.
I've just updated to the fresh released v2.0.2.

I think this can be closed.

Thanks.

**Comment by wtarreau** [comment]
Much appreciated, thank you!
